### PR TITLE
Migration JS AST to Estree

### PR DESCRIPTION
**Pull request by smelukov**

### CODE REVIEW

File: lib/js/scope.js [hunk 1]
```diff
@@ -59,55 +62,68 @@ function process(ast, scope){
...
+
+      this.scopes.push(scope);
     },
-    'name': function(token, scope){
-      var name = token[1];
 
-      if (name == 'null' || name == 'false' || name == 'true')
-        return;
+    Identifier: function(token, scope){
+      // todo implement more efficient solution in a future version
```
Comment (lahmatiy) [inline lib/js/scope.js]: +1

File: lib/js/utils.js [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+    return node;
+  },
+  /**
+   * Create literal from string, number or boolean
+   *
+   * @param {string|number|boolean} value
+   * @returns {?Literal}
+   */
+  createLiteral: function(value){
```
Comment (lahmatiy) [inline lib/js/utils.js]: Optimization needed

File: lib/js/utils.js [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+   * Create identifier
+   *
+   * @param {string} name
+   * @param {boolean} wrap
+   * @returns {Identifier}
+   */
+  createIdentifier: function(name, wrap){
+    return {
+      type: 'Identifier',
+      name: wrap ? ('"' + name + '"') : name
```
Comment (lahmatiy) [inline lib/js/utils.js]: No parenthesis, please

File: lib/js/utils.js [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+    return node;
+  },
+  /**
+   * Prepend node to block or function
+   *
+   * @param {BlockStatement|ForStatement|SwitchStatement|ClassBody} node
+   * @param {Node} prepend
+   */
+  prepend: function(node, prepend){
```
Comment (lahmatiy) [inline lib/js/utils.js:114]: Optimization needed

File: lib/js/utils.js [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+
+    Array.prototype.unshift.apply(node, prepend);
+  },
+  /**
+   * Append node to block or function
+   *
+   * @param {BlockStatement|ForStatement|SwitchStatement|ClassBody} node
+   * @param {Node} append
+   */
+  append: function(node, append){
```
Comment (lahmatiy) [inline lib/js/utils.js:132]: Optimization needed

File: lib/js/utils.js [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+    Array.prototype.push.apply(node, append);
+  },
+  /**
+   * Replace node, but save the reference to the old node object
+   *
+   * @param {Node} oldNode
+   * @param {Node} newNode
+   * @return {Node}
+   */
+  replaceNode: function(oldNode, newNode){
```
Comment (lahmatiy) [inline lib/js/utils.js:151]: Optimization needed

File: lib/js/utils.js [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+  isFunction: function(node){
+    return this.isTypeFunction(node.type);
+  },
+  /**
+   * Is type a function
+   *
+   * @param {string} type node type
+   * @returns {boolean}
+   */
+  isTypeFunction: function(type){
```
Comment (lahmatiy) [inline lib/js/utils.js]:
```
return type === '...' || type === '...' || ...
```

File: lib/js/utils.js [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+  isClass: function(node){
+    return this.isTypeClass(node.type);
+  },
+  /**
+   * Is type a class
+   *
+   * @param {string} type node type
+   * @returns {boolean}
+   */
+  isTypeClass: function(type){
```
Comment (lahmatiy) [inline lib/js/utils.js]:
```
type === '...' || ...
```

File: lib/js/utils.js [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+        return this.cloneArray(el);
+
+      if (this.isObject(el) && el.constructor === Object)
+        return this.deepExtend({}, el);
+
+      return el;
+    }, this);
+  },
+  extend: function(target){
+    var sources = Array.prototype.slice.call(arguments, 1);
```
Comment (lahmatiy) [inline lib/js/utils.js]: Don't clone arguments, it's slow

File: lib/js/walker.js [hunk 10]
```diff
@@ -36,243 +39,59 @@ function ast_walker(){
...
+        }
 
-      case 'object':   // type, props
+        var userFn = user[token.type] || user['*'];
 
-        for (var i = 0, props = token[1], prop; prop = props[i]; i++)
-          walk(prop, 1);
+        if (userFn)
+        {
+          var managerContext = Object.create(walkerContext);
```
Comment (lahmatiy) [inline lib/js/walker.js]: I suppose those 3 lines aren't performant. You create a new object and two bounded functions for _each_ node on walking.

File: package.json [hunk 11]
```diff
@@ -25,9 +25,12 @@
     "test": "mocha"
   },
   "dependencies": {
-    "htmlparser2": "~3.9.2",
+    "htmlparser2": "3.9.1",
```
Comment (lahmatiy) [inline package.json]: Why version is changed?

Comment (smelukov) [reply package.json]: must be some wrong merging
I'll fix it

File: lib/js/walker.js (hunk 10 above)
Comment (smelukov) [reply lib/js/walker.js]: remove it and used a feature of estraverse - [estraverse.VisitorOption.Skip](https://github.com/estools/estraverse/wiki/Usage)

File: lib/js/structure.js [hunk 12]
```diff
@@ -21,7 +21,7 @@ function process(ast, context){
         if (fn.ref_)
           fn = fn.ref_;
 
-        if (fn[0] == 'function' || fn[0] == 'defun')
+        if (['FunctionExpression', 'FunctionDeclaration', 'ArrowFunctionExpression'].indexOf(fn.type) > -1)
```
Comment (lahmatiy) [inline lib/js/structure.js]: fn.type === '...' || ..

File: lib/js/structure.js [hunk 13]
```diff
@@ -31,7 +31,7 @@ function process(ast, context){
             fn.run.call(
               this,
               token,
-              expr[0] == 'dot' || expr[0] == 'name' ? scope.resolve(expr[1]) : null,
+              ['Identifier', 'MemberExpression'].indexOf(expr.type) > -1 ? scope.resolve(expr.object || expr.name) : null,
```
Comment (lahmatiy) [inline lib/js/structure.js]: expr.type === '...' || ...

File: lib/js/utils.js [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+   */
+  createLiteral: function(value){
+    var valueType = typeof value;
+
+    if (valueType === 'string' || valueType === 'number' || valueType === 'boolean')
+    {
+      return {
+        type: 'Literal',
+        value: value,
+        raw: valueType === 'string' ? '"' + value + '"' : String(value)
```
Comment (lahmatiy) [inline lib/js/utils.js]: For string you need something like:
```js
value.replace(/\\/g, '\\\\').replace(/\r/g, '\\r').replace(/\n/g, '\\n').replace(/"/g, '\\"')....
```
I think it's better to use `JSON.stringify()` here

File: lib/js/utils.js [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+   *
+   * @param {BlockStatement|ForStatement|SwitchStatement|ClassBody} node
+   * @param {Node} prepend
+   */
+  prepend: function(node, prepend){
+    var prependTo = this.getBody(node);
+    var prependFrom = prepend;
+    var firstExpr = this.getFirstExpression(prependTo);
+
+    // if the first expression of the body is a function, then prepend nodes to its body
```
Comment (lahmatiy) [inline lib/js/utils.js]: Something strange here. Why we need this kind of magick?

File: lib/js/utils.js [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+  },
+  /**
+   * Create identifiers
+   *
+   * @param {Array<string>} names
+   * @returns {Array<Identifier>}
+   */
+  createIdentifiers: function(names){
+    if (Array.isArray(names))
+      return names.map(function(name){
```
Comment (lahmatiy) [inline lib/js/utils.js]: `.map(this.createIdentifier)`

File: lib/js/utils.js [hunk 17]
```diff
@@ -0,0 +1,384 @@
...
+  /**
+   * Create variable declaration block
+   *
+   * @param {string} kind may be `var`, `let` or `const`
+   * @param {Array<string>|{string:?Node}} variables if object, then keys is names, values is init values
+   * @returns {VariableDeclaration}
+   */
+  variableBlock: function(kind, variables){
+    var block = {
+      type: "VariableDeclaration",
```
Comment (lahmatiy) [inline lib/js/utils.js]: Single quotes please

File: lib/js/walker.js [hunk 18]
```diff
@@ -36,245 +39,60 @@ function ast_walker(){
...
-
-          walkEach(branch[1]);  // branch body
+          var result = userFn.call(walkerContext, token, scope);
+
+          if (token.scope && token.scope !== scope)
+          {
+            scope = token.scope;
+            scopeStack.push(scope);
+          }
+          if(result === walkerContext.SKIP)
```
Comment (lahmatiy) [inline lib/js/walker.js]: `if(` -> `if (`

**Pull request merged**